### PR TITLE
[JUJU-1877] Improve "juju info" channels output

### PR DESCRIPTION
**Pull request by benhoyt**
This updates the "juju info" text format output as per my [proposal doc](https://docs.google.com/document/d/1iehQm0R8w4OUJfakY8UXJCIMt073b8r-FW1ECzXcWmc/edit).

Specifically, it shows the arch and bases where applicable as follows:

- When no --series or --arch is specified, show just the latest revision for each channel (including the architecture for clarity)
- When --series is specified, show one line (revision) per arch
- When --arch is specified, show a comma-separated list of series (there's only one latest revision per arch)
- When both series and arch are specified, show just the latest revision (but filtered by series and arch)

This also changes the YAML and JSON output formats in 3.0, mainly so we can include the multiple revisions per track/risk combination (charms normally have one revision per architecture now, though note that old Charmstore charms can be returned from the CharmHub API as arch="all").

This PR also includes the update to "juju info" YAML/JSON output to show bases as "ubuntu:22.04" rather than "ubuntu/22.04". This should merge cleanly once [this commit](https://github.com/juju/juju/commit/3a279999adbe6b21d6719540e36df4066b3dee1f#diff-2e6e3fd3f525cab09fb3666efde3b00ed439ce2d1bb583d14b9f307d96b12c54) is merged into develop.

However, work to add `--base` and convert non-YAML/JSON output to show bases is left to a subsequent PR.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```sh
$ juju info ubuntu-advantage
name: ubuntu-advantage
...
channels: |
  latest/stable:     33  2022-06-13  (33)  1MB  armhf  xenial, bionic, focal, jammy
  latest/candidate:  ↑
  latest/beta:       ↑
  latest/edge:       38  2022-06-13  (38)  1MB  arm64  xenial, bionic, focal, jammy

$ juju info ubuntu-advantage --series jammy
name: ubuntu-advantage
...
channels: |
  latest/stable:     33  2022-06-13  (33)  1MB  armhf
  latest/candidate:  ↑
  latest/beta:       ↑
  latest/edge:       34  2022-06-13  (34)  1MB  amd64
                     38  2022-06-13  (38)  1MB  arm64
                     37  2022-06-13  (37)  1MB  armhf
                     35  2022-06-13  (35)  1MB  ppc64el
                     36  2022-06-13  (36)  1MB  s390x

$ juju info ubuntu-advantage --arch amd64
name: ubuntu-advantage
...
channels: |
  latest/stable:     16  2022-03-02  (16)  1MB  xenial, bionic, focal
  latest/candidate:  ↑
  latest/beta:       ↑
  latest/edge:       34  2022-06-13  (34)  1MB  xenial, bionic, focal, jammy

$ juju info ubuntu-advantage --series jammy --arch amd64
name: ubuntu-advantage
...
channels: |
  latest/stable:     –
  latest/candidate:  –
  latest/beta:       –
  latest/edge:       34  2022-06-13  (34)  1MB
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1980181

### CODE REVIEW

File: cmd/juju/charmhub/convert.go [hunk 1]
```diff
@@ -139,7 +140,7 @@ func channelSeries(platforms []corecharm.Platform) set.Strings {
 func channelBases(platforms []corecharm.Platform) set.Strings {
 	bases := set.NewStrings()
 	for _, v := range platforms {
-		bases.Add(fmt.Sprintf("%s/%s", v.OS, v.Channel))
+		bases.Add(fmt.Sprintf("%s:%s", v.OS, v.Channel))
```
Comment (benhoyt) [inline cmd/juju/charmhub/convert.go:143]: I'm including this change in here as well. Should merge cleanly when https://github.com/juju/juju/commit/3a279999adbe6b21d6719540e36df4066b3dee1f#diff-2e6e3fd3f525cab09fb3666efde3b00ed439ce2d1bb583d14b9f307d96b12c54 is merged into develop.

File: cmd/juju/charmhub/convert.go [hunk 2]
```diff
@@ -288,36 +289,24 @@ func formatRelationPart(r map[string]charm.Relation) (map[string]string, bool) {
 	return relations, true
 }
 
-type charmMeta struct {
```
Comment (benhoyt) [inline cmd/juju/charmhub/convert.go:291]: This type was no longer used.

File: cmd/juju/charmhub/data.go [hunk 3]
```diff
@@ -43,13 +47,13 @@ type FindResponse struct {
...
 }
 
-type Channel struct {
-	ReleasedAt string   `json:"released-at" yaml:"released-at"`
+type Revision struct {
 	Track      string   `json:"track" yaml:"track"`
 	Risk       string   `json:"risk" yaml:"risk"`
+	Version    string   `json:"version" yaml:"version"`
 	Revision   int      `json:"revision" yaml:"revision"`
+	ReleasedAt string   `json:"released-at" yaml:"released-at"`
```
Comment (benhoyt) [inline cmd/juju/charmhub/data.go:55]: I reordered the fields here for what I think is a slightly nicer order in the YAML output.

File: cmd/juju/charmhub/info_test.go [hunk 4]
```diff
@@ -82,8 +85,110 @@ func (s *infoSuite) TestRunJSON(c *gc.C) {
...
-`)
+	c.Assert(indentJSON(c, cmdtesting.Stdout(ctx)), gc.Equals, `
```
Comment (benhoyt) [inline cmd/juju/charmhub/info_test.go:88]: These tests with the expected JSON on a single line were super-hard to maintain, so I converted the expected value to indented JSON.

File: cmd/juju/charmhub/info_test.go [hunk 5]
```diff
@@ -204,19 +574,20 @@ charm:
...
-    risk: stable
-    revision: 16
-    size: 12042240
-    version: 1.0.3
-    architectures:
-    - amd64
-    bases:
-    - ubuntu/16.04
-    - ubuntu/18.04
+channels:
```
Comment (benhoyt) [inline cmd/juju/charmhub/info_test.go:577]: The diff here gives you an idea of what's actually changed structurally in the YAML/JSON output.

File: tests/suites/deploy/deploy_bundles.sh [hunk 6]
```diff
@@ -104,9 +104,9 @@ run_deploy_exported_charmhub_bundle_with_float_revisions() {
 	juju deploy ${bundle}
 
 	echo "Create telegraf_bundle_without_revisions.yaml with known latest revisions from charmhub"
-	influxdb_rev=$(juju info influxdb --format json | jq -r '."channel-map"."latest/stable".revision')
-	telegraf_rev=$(juju info telegraf --format json | jq -r '."channel-map"."latest/stable".revision')
-	ubuntu_rev=$(juju info ubuntu --format json | jq -r '."channel-map"."latest/stable".revision')
+	influxdb_rev=$(juju info influxdb --format json | jq -r '."channels"."latest"."stable"[0].revision')
+	telegraf_rev=$(juju info telegraf --format json | jq -r '."channels"."latest"."stable"[0].revision')
+	ubuntu_rev=$(juju info ubuntu --format json | jq -r '."channels"."latest"."stable"[0].revision')
```
Comment (benhoyt) [inline tests/suites/deploy/deploy_bundles.sh:109]: This diff shows the changes that are necessary to consumers of `juju info --format=json|yaml`.

File: cmd/juju/charmhub/infowriter_test.go [hunk 7]
```diff
@@ -21,13 +21,47 @@ var _ = gc.Suite(&printInfoSuite{})
...
+tags: app, seven
+subordinate: true
+relations:
+  provides:
+    one: two
+    three: four
+  requires:
+    five: six
+channels: |
+  latest/stable:     1.0.3  2019-12-16  (16)  12MB  amd64  jammy, focal
```
Comment (wallyworld) [inline cmd/juju/charmhub/infowriter_test.go:47]: For status and show-machine / show-application, we're now transitioning away from series to display base in tabular output, ie "ubuntu:22.04" etc.
I wonder if we want to do the same here. Charmhub shows the version numbers in its UX. We should be consistent with that I think. We could use a similar UX, eg "ubuntu: 18.04 20.04 22.04"

**Review by wallyworld — APPROVED** [review]
Nice improvement. I think though we need to transition away from series as we are doing elsewhere in 3.0 and as charmhub already has. Plus coolos uses os version number, so we should just bite the bullet and do it consistently for ubuntu also.

**Pull request merged**